keps/sig-node/2238-liveness-probe-grace-period/README.md

```diff
@@ -0,0 +1,491 @@
+# KEP-2238: Liveness Probe Grace Periods
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Configuration example](#configuration-example)
+  - [User Stories (Optional)](#user-stories-optional)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Liveness probes currently use the `terminationGracePeriodSeconds` on both
+normal shutdown and when probes fail. Hence, if a long termination period is
+set, and a liveness probe fails, a workload will not be promptly restarted
+because it will wait for the full termination period.
+
+I propose adding a new field to liveness probes, `gracePeriodSeconds`. When
```

Comment:
Do we need to say that this field is not applicable to startupProbe and readinessProbe explicitly?

Reply:
I detest the existence of fields in re-used structs that can only be used in some contexts.  If we do this, KEP, can we please create a new struct (stasring with a copy of the old) for those places that need the new fields?

Also - wouldn't startupProbes use this?

Reply:
So I sort of left this intentionally vague as a discussion point :)

I think that readiness probes use a different code path, but could be configured with this value if we wanted it to. I don't know if the bug currently affects startup probes, since they're pretty new.

Reply:
I don't think this field is also needed for the `startupProbe` as it only terminates the container when the probe has never succeeded. So I don't think there is anything to cleanup which would require a long termination period.
Maybe I am wrong for some use cases where the `startupProbe` is used to cover a set of DB scripts?

Reply:
I think `startupProbes` are affected. Upon probe failure, both liveness and startup probes queue the container to be killed: https://github.com/kubernetes/kubernetes/blob/e414d4e5c2ab85eada7eb8e80a363658e199968b/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L631-L655 This is presumably why the `terminationGracePeriodSeconds` from the pod spec is used for termination.

Readiness probes go through a different code path (indicated by [this comment](https://github.com/kubernetes/kubernetes/blob/e414d4e5c2ab85eada7eb8e80a363658e199968b/pkg/kubelet/prober/prober_manager.go#L55)) although I haven't figured out the exact difference yet.

Reply:
I've updated the KEP to explicitly call out liveness/startup probes, and not readiness.

Reply:
(pasting this for reference here, this is where the grace termination periods are currently set upon container kill: https://github.com/kubernetes/kubernetes/blob/e414d4e5c2ab85eada7eb8e80a363658e199968b/pkg/kubelet/kuberuntime/kuberuntime_container.go#L604-L611)